File: java/Queues/CircularQueue.java
package Queues;

import definitions.Queues;

import java.util.Arrays;

public class CircularQueue<T> implements Queues<T> {

    private Object[] arr;
    private int front, rear, count;

    public CircularQueue(int capacity) {
        arr = new Object[capacity];
        init();
    }

    private void init() {
        count = 0;
        front = -1;
        rear = -1;
    }

    @Override
    public boolean isEmpty() {
        return front == -1 && rear == -1;
    }

    @Override
    public void clear() {
        Arrays.fill(arr, null);
        count = 0;
    }

    private boolean isFull() {
        return (rear + 1) % arr.length == front;
    }

    @Override
    public void add(T val) {
        if (isFull()) return;
        if (isEmpty()) {
            front = 0;
            rear = 0;
        } else {
            rear = (rear + 1) % arr.length;
        }
        arr[rear] = val;
        count++;
    }

    @Override
    public T remove() {
        if (isEmpty()) return null;
        var val = (T) arr[front];
        arr[front] = null;
        front = (front + 1) % arr.length;
        if (isEmpty()) init();
        else count--;
        return val;
    }

    @Override
    public int size() {
        return count;
    }

    @Override
    public T peek() {
        return (T) arr[front];
    }

    @Override
    public String toString() {
        return Arrays.toString(arr);
    }

}

File: typescript/src/queues/circular-queue.ts
export default class CircularQueue<T> {
  private arr: T[];
  private arrSize: number;
  private count: number;

  private rear: number;
  private front: number;

  constructor(size: number) {
    this.arrSize = size;
    this.arr = new Array(size);
    this.count = 0;
    this.rear = -1;
    this.front = -1;
  }

  private isEmpty() {
    return this.rear == -1 && this.front == -1;
  }

  private isFull() {
    return (this.rear + 1) % this.arrSize == this.front;
  }

  public clear() {
    this.arr = new Array(this.arrSize);
    this.count = 0;
    this.front = -1;
    this.rear = -1;
  }

  public add(val: T) {
    if (this.isFull()) return;
    if (this.isEmpty()) {
      this.arr[this.rear++] = val;
      this.front++;
      this.count++;
      return;
    }
    this.rear = (this.rear + 1) % this.arrSize;
    this.arr[this.rear] = val;
    this.count++;
  }

  public remove() {
    if (this.isEmpty()) return null;
    const val = this.arr[this.front];
    this.front = (this.front + 1) % this.arrSize;
    if (this.isEmpty()) {
      this.clear();
    } else this.count--;
    return val;
  }

  public peek() {
    if (this.isEmpty()) return null;
    return this.arr[this.front];
  }

  public get size() {
    return this.arrSize;
  }
}
